contrib/masterfiles-stage/common.sh

```diff
@@ -12,175 +12,250 @@ if [ ! "$debug_mode" == "true" ]; then
+  # 3. sets appropriate permissions on the policy set,
+  # 4. validates the policy set using cf-promises,
+  # 5. moves the temp dir policy set into the given deploy dir,
+  #    avoiding triggering policy updates unnecessarily
+  #    by comparing the cf_promises_validated flag file.
+  #    (See long comment at end of function def.)
+
+  # Ensure absolute pathname is given
+  [ "${1:0:1}" = / ] ||
+    error_exit "You must specify absolute pathnames in channel_config: '$1'"
+  mkdir -p "$(dirname "$1")" || error_exit "Failed to mkdir -p dirname $1"
+    # We don't mkdir $1 directly, just its parent dir if that doesn't exist.
+
+  ########################## 1. CREATE EMPTY TEMP DIR
+  # Put staging dir right next to deploy dir to ensure it's on same filesystem
+  local temp_stage
+  temp_stage="$(mktemp -d --tmpdir="$(dirname "$1")" )"
+
+  ########################## 2. CHECKOUT INTO TEMP DIR
+  # The '^0' at the end of the refspec
+  # populates HEAD with the SHA of the commit
+  # rather than potentially using a git branch name.
+  # Also see http://stackoverflow.com/a/13293010/5419599
+  # and https://github.com/cfengine/core/pull/2465#issuecomment-173656475
+  git --git-dir="${local_mirrored_repo}" --work-tree="${temp_stage}" checkout -q -f "${2}^0" ||
+    error_exit "Failed to checkout '$2' from '${local_mirrored_repo}'"
+  # Grab HEAD so it can be used to populate cf_promises_release_id
+  mkdir -p "${temp_stage}/.git"
+  cp "${local_mirrored_repo}/HEAD" "${temp_stage}/.git/"
+
+  ########################## 3. SET PERMISSIONS ON POLICY SET
+  chown -R root:root "${temp_stage}" || error_exit "Unable to chown '${temp_stage}'"
+  find "${temp_stage}" \( -type f -exec chmod 600 {} + \) -o \
+                       \( -type d -exec chmod 700 {} + \)
+
+  ########################## 4. VALIDATE POLICY SET
+  /var/cfengine/bin/cf-promises -T "${temp_stage}" &&
+  /var/cfengine/bin/cf-promises -cf "${temp_stage}/update.cf" ||
+  error_exit "Update policy staged in ${temp_stage} could not be validated, aborting."
+
+  ########################## 5. ROLL OUT POLICY SET FROM TEMP DIR TO DEPLOY DIR
+  if ! [ -d "$1" ] ; then
+    # deploy dir doesn't exist yet
+    mv "${temp_stage}" "$1" || error_exit "Failed to mv $temp_stage to $1."
   else
-    error_exit "The staged policies in ${STAGING_DIR} could not be validated, aborting."
+    if /usr/bin/cmp -s "${temp_stage}/cf_promises_release_id" \
+                                 "${1}/cf_promises_release_id" ; then
+      # release id is the same in stage and deploy dir
+      # so prevent triggering update on hosts by keeping old "validated" flag file
+      cp -a "${1}/cf_promises_validated" "${temp_stage}/"
+    fi
+    local third_dir
+    third_dir="$(mktemp -d --tmpdir="$(dirname "$1")" )"
+    mv "${1}" "${third_dir}"  || error_exit "Can't mv ${1} to ${third_dir}"
+    mv "${temp_stage}" "${1}"          || error_exit "Can't mv ${temp_stage} to ${1}"
+    rm -rf "${third_dir}"
   fi
```

Comment:
So we could skip the deployment if the release id is the same to avoid the "potentially" unncessary churn. I say potentially because we only operate on partial knowledge of how different the currently deployed policy set is from the staged one (only looking at the release_id).

I dislike that assumption, but I don't have any great ideas to do further validation without it getting potentially expensive. At least move operations on the same file system are fast and not cpu intensive. A recursive diff or hash could get expensive.